src/calls.c

```diff
@@ -550,267 +568,88 @@ Obj DoProf0args (
-    SET_TIME_WITH_PROF( prof, SyTime() - timeElse );
-    SET_STOR_WITH_PROF( prof, SizeAllBags - storElse );
-
-    /* time and storage spent by this invocation of this function          */
-    timeCurr = SyTime() - TimeDone - timeCurr;
-    SET_TIME_WOUT_PROF( prof, TIME_WOUT_PROF(prof) + timeCurr );
-    TimeDone += timeCurr;
-    storCurr = SizeAllBags - StorDone - storCurr;
-    SET_STOR_WOUT_PROF( prof, STOR_WOUT_PROF(prof) + storCurr );
-    StorDone += storCurr;
-
-    /* return the result from the function                                 */
-    return result;
+    return DoProfNNNargs(self, 3, arg1, arg2, arg3, 0, 0, 0);
 }
 
 
 /****************************************************************************
 **
 *F  DoProf4args( <self>, <arg1>, ... )  . profile a function with 4 arguments
 */
-Obj DoProf4args (
+static Obj DoProf4args (
     Obj                 self,
     Obj                 arg1,
     Obj                 arg2,
     Obj                 arg3,
     Obj                 arg4 )
 {
-    Obj                 result;         /* value of function call, result  */
-    Obj                 prof;           /* profiling bag                   */
-    UInt                timeElse;       /* time    spent elsewhere         */
-    UInt                timeCurr;       /* time    spent in current funcs. */
-    UInt8               storElse;       /* storage spent elsewhere         */
-    UInt8               storCurr;       /* storage spent in current funcs. */ 
-
-    /* get the profiling bag                                               */
-    prof = PROF_FUNC( PROF_FUNC( self ) );
-
-    /* time and storage spent so far while this function what not active   */
-    timeElse = SyTime() - TIME_WITH_PROF(prof);
-    storElse = SizeAllBags - STOR_WITH_PROF(prof);
-
-    /* time and storage spent so far by all currently suspended functions  */
-    timeCurr = SyTime() - TimeDone;
-    storCurr = SizeAllBags - StorDone;
-
-    /* call the real function                                              */
-    result = CALL_4ARGS_PROF( self, arg1, arg2, arg3, arg4 );
-
-    /* number of invocation of this function                               */
-    SET_COUNT_PROF( prof, COUNT_PROF(prof) + 1 );
-
-    /* time and storage spent in this function and its children            */
-    SET_TIME_WITH_PROF( prof, SyTime() - timeElse );
-    SET_STOR_WITH_PROF( prof, SizeAllBags - storElse );
-
-    /* time and storage spent by this invocation of this function          */
-    timeCurr = SyTime() - TimeDone - timeCurr;
-    SET_TIME_WOUT_PROF( prof, TIME_WOUT_PROF(prof) + timeCurr );
-    TimeDone += timeCurr;
-    storCurr = SizeAllBags - StorDone - storCurr;
-    SET_STOR_WOUT_PROF( prof, STOR_WOUT_PROF(prof) + storCurr );
-    StorDone += storCurr;
-
-    /* return the result from the function                                 */
-    return result;
+    return DoProfNNNargs(self, 3, arg1, arg2, arg3, arg4, 0, 0);
```

Comment:
`self, 4`? Also, this passed tests so we might want some boring tests just to make sure all these different length functions work properly.

Reply:
Ouch, bad case of copy & paste. Actually, the tests (some which were rather recently added) showed me already another bug I introduced while doing this.

Unfortunately, adding sensible tests for function profiling is rather annoying.

Reply:
Now fixed. In my defense, at least now it was very easy to spot this; we had similar bugs in very similar code for a long time, which were only resolved last year, and they were much harder to spot, as things were spread out over pages and pages of code.

That's of course no excuse for wrong code, but it shows that reducing the duplication makes it easier to spot bugs :-)